File: dynamic/Solution343.java
class Solution343 {
    public int integerBreak(int n) {
    	if(n <= 2){
    		return 1;
    	}
    	int[] dp = new int[n + 1];
    	dp[1] = 1;
    	dp[2] = 1;
    	for(int i = 3; i <= n; i++){
    		int max = -1;
    		for(int j = 1; j < i; j++){
    			int cur = Math.max(dp[j], j) * Math.max(dp[i - j], i - j);
    			max = Math.max(max, cur);
    		}
    		dp[i] = max;
    	}
    	return dp[n];
    }

    public static void main(String[] args){
    	Solution343 s = new Solution343();
    	System.out.println(s.integerBreak(4));
    }
}

File: C++/Solution343.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<stack>
using namespace std;

class Solution343 {
public:
    int integerBreak(int n) {
        vector<int> vec(n + 1);
        vec[1] = 1;
        vec[2] = 1;
        for(int j = 3; j <= n; j++){
            int res = 0;
            for(int i = 1; i < j; i++){
                int cur = max(vec[i], i) * max(j - i, vec[j - i]);
                res = max(res, cur);
            }
            vec[j] = res;
        }
        return vec[n];
    }
};